examples/graph/mpnn-molecular-graphs.py

```diff
@@ -0,0 +1,723 @@
+The SMILES string is a compact encoding which, for smaller molecules, is relatively
+human-readable. Encoding molecules as a string both alleviates and facilitates database
+and/or web searching of a given molecule. RDKit uses algorithms to
+accurately transform a given SMILES to a molecule object, which can then
+be used to compute a great number of molecular properties/features.
+
+Notice, RDKit is commonly installed via [Conda](https://www.rdkit.org/docs/Install.html).
+However, thanks to
+[rdkit_platform_wheels](https://github.com/kuelumbus/rdkit_platform_wheels), rdkit
+can now (for the sake of this tutorial) be installed easily via pip, as follows:
+
+```
+pip -q install rdkit-pypi
+```
+
+And for easy and efficient reading of csv files and visualization, the below needs to be
+installed:
+
+```
+pip -q install pandas
+pip -q install Pillow
+pip -q install matplotlib
+pip -q install pydot
+sudo apt-get -qq install graphviz
+```
+"""
+
+"""
+### Import packages
+"""
+
+import tensorflow as tf
+from tensorflow import keras
+import numpy as np
+import pandas as pd
+import matplotlib.pyplot as plt
+import warnings
+from rdkit import Chem, RDLogger
+from rdkit.Chem.Draw import IPythonConsole, MolsToGridImage
+import logging
+
+tf.get_logger().setLevel(logging.ERROR)
+warnings.filterwarnings("ignore")
+RDLogger.DisableLog("rdApp.*")
+
+np.random.seed(42)
+tf.random.set_seed(42)
+
+"""
+## Dataset
+
+Information about the dataset can be found in
+[A Bayesian Approach to in Silico Blood-Brain Barrier Penetration Modeling](https://pubs.acs.org/doi/10.1021/ci300124c)
+and [MoleculeNet: A Benchmark for Molecular Machine Learning](https://arxiv.org/abs/1703.00564).
+The dataset will be downloaded from [MoleculeNet.ai](http://moleculenet.ai/datasets-1).
+
+### About
+The dataset contains **2,050** molecules; each molecule come with a **name**, **label**
+and **SMILES** string.
+
+The blood-brain barrier (BBB) is a membrane separating the blood from the brain
+extracellular fluid, hence blocking out most drugs (molecules) from reaching
+the brain. Because of this, the BBBP has been important to study for the development of
+new drugs that aim to target the central nervous system. The labels for this
+data set are binary (1 or 0) and indicate the permeability of the molecules.
+"""
+
+csv_path = keras.utils.get_file(
+    "BBBP.csv", "https://deepchemdata.s3-us-west-1.amazonaws.com/datasets/BBBP.csv"
+)
+
+df = pd.read_csv(csv_path, usecols=[1, 2, 3])
+df.iloc[96:104]
+
+"""
+### Define helper functions
+
+The functions below will help us generate graph features, namely, atom and bond features.
+Thanks to RDKit this can be done conventiently with only a few lines of code. Feel free
+to look up additional features that can be generated with
+[RDKit](https://www.rdkit.org/docs/Cookbook.html).
+"""
+
+
+def onehot_encode(x, allowable_set):
+    return list(map(lambda s: float(x == s), allowable_set))
```

Comment:
This will iterate over the elements of the list every time you encode something. It's quite inefficient. Consider refactoring to use dicts/sets.

It seems you want to specify your allowed values as static lists. This is fine. But such lists will have to be turned into hashtables, and you only want to do this conversion once. So it sounds like you're going to want to use a class.

```python
class Index:
  def __init__(self, allowable_values):
    self.allowable_values = sorted(list(allowable_values))
    self.value_index = dict(zip(self.allowable_values, range(len(self.allowable_values))))

  def encode(self, value):
      if value not in self.value_index:
        raise ValueError(...)
      output = np.zeros((len(self.value_index),))
      output[self.value_index[value]] = 1.
      return output

atom_index = Index({"B", "Br", "C", "Ca", "Cl", "F", "I", "N", "Na", "O", "P", "S"})
output = atom_index.encode(atom.GetSymbol())
```

For practicality you may want to merge everything into a single big class that's aware of the atom semantics.

Reply:
Thank's a lot for the feedback, much appreciated. I'll look into it!